libevent.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2020 Abdulrahman Semrie <hsamireh@gmail.com>
+ * Copyright (C) 2020 Aleix Conchillo Flaqué <aconchillo@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 3 of
+ * the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA
+ */
+
+
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <event2/event.h>
+#include <libguile.h>
+
+struct event_data
+{
+  int fd;
+  short event;
+};
+
+struct wait_data
+{
+  int rv;
+  struct event_data *events;
+  int maxevents;
+};
+
+struct loop_data
+{
+  struct event_base *base;
+  int64_t timeout;
+};
+
+static void
+free_evb (void* ev)
+{
+  struct event_base *base = ev;
+  event_base_free (base);
+}
+
+static void
+cb_func (evutil_socket_t fd, short what, void *arg)
+{
+  struct wait_data* data = arg;
+  int rv = data->rv;
+
+  if (rv == data->maxevents)
+    {
+      fprintf (stderr, "max events fired on [%d], ignoring.\n", fd);
+      return;
+    }
+
+  struct event_data ev_data = { fd, what };
+  int sz = sizeof (struct event_data);
+  memcpy (data->events + (rv * sz), &ev_data, sz);
+
+  data->rv += 1;
+}
+
+static SCM
+scm_primitive_event_wake (SCM wakefd)
+#define FUNC_NAME "primitive-event-wake"
+{
+  int c_fd;
+  char zero = 0;
+
+  c_fd = scm_to_int (wakefd);
+
+  if (write (c_fd, &zero, 1) <= 0 && errno != EWOULDBLOCK && errno != EAGAIN)
+    SCM_SYSERROR;
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_create_event_base (SCM eventsv)
+#define FUNC_NAME "primitive-create-event-base"
+{
+  struct event_base *base;
+  struct wait_data *data;
+
+  if ((base = event_base_new ()) == NULL)
+    scm_syserror ("couldn't instantiate event_base");
+
+  data = (struct wait_data *) scm_gc_malloc (sizeof (struct wait_data),
+                                             "wait_data");
+  data->rv = 0;
+  data->events = (struct event_data *) SCM_BYTEVECTOR_CONTENTS (eventsv);
```

Comment:
@wingo moving the conversation to the right PR. In this old comment https://github.com/wingo/fibers/pull/42#pullrequestreview-614124016 you mention `eventsv` should be kept in `data` for GC reasons. I believe this makes sense. And if I understand correctly, that would mean keeping a reference to the `SCM` object, right?

I was looking at the `epoll.c` code and I believe it suffers from the same, so I guess that would be necessary there as well?

cc: @emixa-d @civodul

Reply:
Thanks! I will leave it like it is, for now.